### ISSUE TITLE
Placeholders removed for accessibility

### ISSUE THREAD

**Issue by skechele** [shopware/platform#14442]
### Shopware Version

6.7

### Affected area / extension

Platform(Default)

### Actual behaviour

In Shopware 6.7 placeholders were removed from the registration / address forms. Support said it was accessibility reasons. 
But all the reasons they gave only mattered if only placeholders were used. If placeholders in combination with labels are used (like shopware did previous), none of the reasons mattered. 

I couldn't find a any blog/article etc. to the reasons why placeholders were removed.
In my opinion just removing placeholder does not increase accessibility but decreases it. 

Answer from support: 
> Warum ist das so bzw. warum wirken sich Placeholder-Texte negativ auf die Barrierefreiheit aus?
> 
> Verlust der Eingabeorientierung: Da der Platzhaltertext beim Tippen ausgeblendet wird, fehlt Nutzer:innen während der Eingabe eine dauerhafte Orientierung darüber, welche Information erwartet wird. Das kann insbesondere für Personen mit kognitiven Einschränkungen problematisch sein.
> 
> Unzureichende Wahrnehmbarkeit: Placeholder sind häufig in einem sehr hellen Grauton dargestellt und erfüllen dadurch oft nicht die erforderlichen Kontrastwerte. Zudem können sie visuell eher wie dekorativer Text wirken und nicht als relevante Eingabeanweisung erkannt werden.
> 
> Keine verlässliche Feldbeschriftung: Platzhalter stellen keine gleichwertige, programmatisch erkennbare Beschriftung dar. Screenreader erfassen sie nicht zuverlässig als Label, wodurch insbesondere Nutzer:innen assistiver Technologien die Zuordnung und den Kontext von Formularfeldern verlieren.

### Expected behaviour

Enable the placeholders on forms again to give additional context what values are expected in the form. 

Or at least provide a feature flag to enable placeholders. Extending every twig file to add placeholders should not be thing we have to do. 

### How to reproduce

Install Shopware 6.7 got the registration page or the the first page in the order flow.

**Comment by ravenokko** [comment]
Hi @skechele and thank you for reaching out. 

> TL;DR: Placeholders place a burden on developers to implement correctly and place a burden on people with cognitive, motor, or visual disabilities even when done right.

Here is an [in-depth article about placeholders from the NN/Group](https://www.nngroup.com/articles/form-design-placeholders/) which clarifies all the reasons why adding a label to a form with a placeholder doesn't solve all the problems placeholders create. The primary points are:

- it adds cognitive overhead (you have to remember what the placeholder said when it disappears)
- people can mistake a field with placeholder text as already filled in (which assumes they didn't overlook it - empty fields are easier to visually spot)
- it's annoying for keyboard users who, when they tab to the form field, aren't looking ahead to read the placeholder text (at which point, the cursor is already set and the placeholder is gone)

The WCAG form instructions tutorial mentions that [browser defaults often do not meet correct contrast levels](https://www.w3.org/WAI/tutorials/forms/instructions/#placeholder-text). So developers also have to fix that, which assumes the browser even lets you easily override the contrast. The problem then is, as accessibility expert Sheri Byrne-Haber notes, [many people using assistive technology still use older browser and OS versions](https://www.sheribyrnehaber.com/accessibility-best-practices-for-screenreader-testing/) (see: Best Practice # 4) due to highly customised tech stacks, economic reasons, or both.  [The Web AIM Survey #10](https://webaim.org/projects/screenreadersurvey10/#browsers) confirms there are people using assistive technology still using IE in 2024 which (along with Opera Mini) [doesn't support `css ::placeholder`](https://caniuse.com/css-placeholder), at all. 

And while these days, the most popular screen readers do seem to better recognize placeholders, there are still [factors that can cause screen reader users to miss or confuse them](https://blog.lipsumhub.com/do-screen-readers-read-placeholder-text/#toc-5). All of which places an additional burden on developers to correctly implement them (plus any styling workarounds) not to mention testing, maintenance and monitoring for regressions. 

The NN/Group recommendation is to remove all placeholders and instead have labels and hint text that are always visible. Between their research, and the advice/experience of accessibility experts, placeholders generally seem to cause more problems than they solve, which is why we have eliminated them in favor of more proven UX.

**Comment by tobiasberge** [comment]
@skechele Hi 👋, thank you for your feedback. It is correct that the mentioned accessibility concerns with placeholders mostly matter when placeholders are used as labels / instead of labels. But the contrast and "invisible when typing" concerns are still valid.

The placeholders were removed from most fields where the information was basically redundant and if anything, only added more "noise". It mostly said the same as the label. A screenreader is aware that it is an input field, so we don't need to tell that you can "enter" something.

```
# Previously:

Email address:
[Enter Email address...]

Password:
[Enter Password..._____]

# Now:

Email address:
[______________________]

Password:
[______________________]

```

And we already have a proper label that describes the field. A placeholder should only be used to give additional, useful information or example inputs like "e.g. John Doe..." etc. And even for "helptext", there is a lot of debate that placeholders are not a good choice. A normal text next to the field linked with `aria-describedby` would be preferable.

* https://www.w3.org/WAI/tutorials/forms/instructions/
* https://www.deque.com/blog/accessible-forms-the-problem-with-placeholders/
* https://www.nngroup.com/articles/form-design-placeholders/

Have you used the translations of the placeholders to provide special information for specific fields? Imo it would not make sense to just add the previous redundant placeholders back that say the same as the label. However, if there is a real use-case to modify/use the placeholders we could think of implementing something for this as an option. Maybe you can elaborate?

**Comment by skechele** [comment]
Thank you both very much for detailed explanations. I know you are not going to revert your decision because of 1 issue, but I hope you give us the ability to turn them back on, or a easier way to set them ourselves. 

Our internal research has shown, that input times were decreasing even with "simple" placeholders like: 

```
Email address:
[Enter Email address...]
```
_We do not provide services for elderly people, but far from IT / technical people_ 

Modifying the inputs is a bit annoying. For example https://github.com/shopware/shopware/blob/v6.7.6.2/src/Storefront/Resources/views/storefront/component/address/field/address-city-field.html.twig
```
{% block component_address_field_city %}
    {% if formViolations.getViolations('/city') is not empty %}
        {% set violationPath = '/city' %}
    {% elseif formViolations.getViolations("/#{prefix}/city") is not empty %}
        {% set violationPath = "/#{prefix}/city" %}
    {% endif %}

    {% set cityAutocomplete = 'address-level2' %}

    {% if prefix == 'shippingAddress' %}
        {% set cityAutocomplete = 'shipping address-level2' %}
    {% elseif prefix == 'billingAddress' %}
        {% set cityAutocomplete = 'billing address-level2' %}
    {% endif %}

    {% sw_include '@Storefront/storefront/component/form/form-input.html.twig' with {
        label: 'address.cityLabel'|trans|sw_sanitize,
        id: idPrefix ~ prefix ~ 'AddressCity',
        name: prefix ~ '[city]',
        value: data.get('city'),
        autocomplete: cityAutocomplete,
        violationPath: violationPath,
        validationRules: 'required',
        additionalClass: additionalClass ?? 'col-md-3 col-8',
    } %}
{% endblock %}
```
Everything has to be duplicated to add 1 placeholder line, which adds a lot of work every update, especially if there a lot of changes. I believe shopware 6.5 and 6.7 had a lot of changes, so changes to these files happen quite frequently.

**Comment by 8mylez-dg** [comment]
@skechele instead of overwriting the whole block, you could also just the extend the corresponding templates and use `{{ parent() }}`:

```twig
{% sw_extends "@Storefront/storefront/component/address/field/address-city-field.html.twig" %}

{% block component_address_field_city %}
    {% set placeholder = "address.cityPlaceholder"|trans %}

    {{ parent() }}
{% endblock %}
```

`placeholder` then gets passed to the form input template and works as expected.

I see your point, though and would also suggest to give us at least the option to turn them on via flag or setting.

**Comment by skechele** [comment]
@8mylez-dg thanks for the reminder, because every attribute was explicitly passed, I forgot that values get handed down passively as well.
